Create Bets

YarekTenko opened the pull request:
Dependent on #28
Allows logged in users to create bets by specifying their text and expiry date.
Created bets can be viewed separately.
All bets will be available through /bets page, where user will be able to specify search criteria to filter through the database.
Ongoing (active bets with status set to 1) will be displayed on the main page

LeonTenko committed:
Add DateTimePicker JQuery UI

LeonTenko committed:
Define create method for BetController

LeonTenko committed:
Automatically log in signed-up users

jacquesporveau commented on app/views/bets/new.html.erb:
```diff
@@ -0,0 +1,30 @@
+
+  <% else %>
+    <h2>You are not logged in </h2>
+<% end %>
```
Identation on the `end`

Also: curious why we aren't using a Rails form_for tag?

YarekTenko replied:
Good catch, thanks! 
Just feels like doing it manually gives me a bit more control over the style, but I might be mistaken here. Is it considered a better idea to use form_for in situations like that?

jacquesporveau replied:
It's considered to be best practices. It's kind of like buying a state of the art chainsaw and then using a handsaw instead.
I think some of the philosophy behind Rails is "lets make things easy". By using the form for tag you are using the Rails magic.

jacquesporveau commented on app/views/layouts/application.html.erb:
```diff
@@ -9,6 +9,10 @@
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-datetimepicker/2.5.4/build/jquery.datetimepicker.full.min.js"></script>
+
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-datetimepicker/2.5.4/build/jquery.datetimepicker.min.css" />
+    <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
```
Looks like Rails is loading the other js and css in the head tag. Putting these in the body seems a bit odd.

YarekTenko replied:
Thank you very much for all the feedback, I'll get to fixing it!
Yeah, I actually tried putting it in the head tag but it wouldn't load for some reason. Maybe I did something wrong, I'll try again.

jacquesporveau commented on app/views/pages/index.html.erb:
```diff
@@ -1,14 +1,55 @@
+      <div class="col col-lg-4">
+      </div>
+      <div class="col col-lg-3">
+      </div>
```
Up to you for style preferences but I think we could single line these last couple lines
`<div class="col col-lg-3></div>` etc.

jacquesporveau commented on app/views/pages/index.html.erb:
```diff
@@ -1,14 +1,55 @@
+      </div>
+      <div class="col col-lg-1">
+        <div class="dropdown">
+            <img class="" type="button" data-toggle="dropdown" src="http://fuuse.net/wp-content/uploads/2016/02/avatar-placeholder.png" style="width: 100%;">
```
4 space indent and not sure we aren't using the Rails image_tag?
http://apidock.com/rails/ActionView/Helpers/AssetTagHelper/image_tag
Also noticing you have a class attr with no data. Not sure if needed?

jacquesporveau commented on app/views/pages/index.html.erb:
```diff
@@ -1,14 +1,55 @@
+          <tr>
+            <th class="col-md-2">
+              Created By
+            </th>
```
You can one line these all if you want. Up to you but I think it looks better.

jacquesporveau commented on app/views/pages/index.html.erb:
```diff
@@ -1,14 +1,55 @@
+        <h2>Welcome, stranger!</h2><br>
+      </div>
+    </div>
       <div class="row">
```
Not part of the PR but your identation is off here too.

jacquesporveau reviewed (commented):
Looking good! Left some general feedback. Most of it is optional.

isaacfreeman commented on app/views/bets/new.html.erb:
```diff
@@ -0,0 +1,30 @@
+<div class="container">
+  <% if current_user %>
+    <h2> Creating a new bet </h2>
+    <form action="<% create_bet_path %>" method="post" id="create-bet-form">
```
In situations like this where there's no instance to to use with `form_for`, Rails provides `form_tag` and corresponding methods like `text_field_tag`, `hidden_field_tag` etc.

isaacfreeman commented on app/views/pages/index.html.erb:
```diff
@@ -1,21 +1,52 @@
+      </div>
+      <div class="col col-lg-4"></div>
+      <div class="col col-lg-3"></div>
+      <div class="col col-lg-1">
```
Can you use a Bootstrap offset class here to replace the empty tags above?

isaacfreeman commented on app/views/pages/index.html.erb:
```diff
@@ -1,21 +1,52 @@
+      <div class="col col-lg-3"></div>
+      <div class="col col-lg-1">
+        <div class="dropdown">
+          <img type="button" data-toggle="dropdown" src="http://fuuse.net/wp-content/uploads/2016/02/avatar-placeholder.png" style="width: 100%;">
```
It looks like this image represents a page control. It's probably best to use a `<button>` here with the image specified in the CSS, and keep `<img>` for images that are content rather than controls.

qr8r commented on config/routes.rb:
```diff
@@ -8,5 +8,7 @@
 
   get '/signup', to: 'users#new', as: :new_user_path
 
+  get  '/create_bet', to: 'bets#new', as: :new_bet
```
You should be using the `resources` syntax in order to create RESTful routes. 

http://guides.rubyonrails.org/routing.html

qr8r commented on app/controllers/bets_controller.rb:
```diff
@@ -1,5 +1,22 @@
 class BetsController < ApplicationController
-  def new_bet
+  def new
```
Should be squashed

qr8r commented on app/views/pages/index.html.erb:
```diff
@@ -1,21 +1,52 @@
+
+    <div class="row">
+      <div class="col col-md-4">
+        <h2>Ongoing bets</h2>
```
This content would be better suited for a bet/index page now

YarekTenko replied:
My idea was to have all the current ongoing (active) bets (sorted by date) listed right on the front page and all the bets ever created on bet/index (with the ability to search/filter/etc.). Would it make sense to leave it here for now in this case?

qr8r commented on app/views/pages/index.html.erb:
```diff
@@ -1,21 +1,52 @@
+        <div class="dropdown">
+          <img type="button" data-toggle="dropdown" src="http://fuuse.net/wp-content/uploads/2016/02/avatar-placeholder.png" style="width: 100%;">
+          <ul class="dropdown-menu">
+            <li><%= link_to 'Log Out', session_path, method: :delete %></li>
```
Since this is content that is likely shared across many pages, It should be extracted into a partial and rendered in the application layout

qr8r commented on app/controllers/bets_controller.rb:
```diff
@@ -1,5 +1,22 @@
+      redirect_to root_path
+    else
+      flash[:danger] = 'Could not create bet'
+      redirect_to create_bet_path
```
Should render the create page so that the form fields stay populated with the values the user tried to use.

qr8r commented on app/controllers/bets_controller.rb:
```diff
@@ -1,5 +1,22 @@
+    @bet = Bet.new(bet_params)
+    if @bet.save
+      flash[:success] = 'Bet has been created'
+      redirect_to root_path
```
Should redirect to the Bet#show page
Same with update when you get there. 

After you create something, as a user, dont you want to see the thing you created?

qr8r commented on app/views/bets/new.html.erb:
```diff
@@ -0,0 +1,30 @@
+    <h2> Creating a new bet </h2>
+    <form action="<% create_bet_path %>" method="post" id="create-bet-form">
+      <div>
+        <input type="hidden" value="<%= form_authenticity_token %>" name="authenticity_token"/>
```
You might want to stick to the rails form helpers. They provide a lot of begin the scenes help for you. 

Once example, is that if you try and edit an existing record. All of the records fields will be pre populated in the form fields
http://guides.rubyonrails.org/form_helpers.html

qr8r commented on config/routes.rb:
```diff
@@ -9,6 +9,7 @@
   get '/signup', to: 'users#new', as: :new_user_path
 
   get  '/create_bet', to: 'bets#new', as: :new_bet
+  post '/create_bet', to: 'bets#create', as: :create_bet
```
use the rails `:resources` syntax. Explicitly defining routes with 'get' etc should be exceptional and used only when absolutely needed

qr8r reviewed (changes requested):
I would like this PR to include a Bet index page and a Bet show page.

They dont need to be fancy or styled. Just present and functional

LeonTenko committed:
Define index method for BetsController

YarekTenko commented:
Thanks for the feedback, everybody, hope I fixed most of the problems this time.

qr8r reviewed (changes requested):
This PR breaks the site

qr8r commented on app/assets/stylesheets/application.scss:
```diff
@@ -20,3 +20,4 @@
 @import "bootstrap";
 @import "font-awesome";
 @import "session-access";
+@import "user-bar";
```
This commit needs a commit message

Adding a controller action and styling page elements should be two separate commits. 
Remember a commit is the smallest piece of coherent and functional code you can put together.

Adding a controller action and Adding a top navigation bar are unrelated and should not be in the same PR, let alone the same commit

qr8r commented on app/views/pages/index.html.erb:
```diff
@@ -1,15 +1,20 @@
+        <h2>Ongoing bets</h2>
+      </div>
+    </div>
+    <%= render 'bets/display_bets' %>
```
This is beyond the scope of the story. 

The story asks only for a page you can visit to create a new bet

qr8r commented on config/routes.rb:
```diff
@@ -8,5 +8,6 @@
 
   get '/signup', to: 'users#new', as: :new_user_path
 
+  resources :bets
```
Are all bet routes available to you at this point?

qr8r commented on app/controllers/sessions_controller.rb:
```diff
@@ -5,7 +5,7 @@ def create
     user = User.find_by(email: params[:session][:email])
     if user && user.authenticate(params[:session][:password])
```
^^ Commit message is empty 

Why are you making this change?

qr8r commented on app/assets/javascripts/application.js:
```diff
@@ -14,5 +14,6 @@
 //= require tether
```
This is a good sized commit

LeonTenko committed:
Add BetsController for Bet

-Added blank new, show and index .html.erb files
-Added resources :bets in config/routes.rb

LeonTenko committed:
Add navigation bar

-app/views/shared/_user_bar.html.erb contains the navigation bar that
will be shared amongs all pages for the logged in user.

LeonTenko committed:
Define show method for BetsController

qr8r commented on app/views/layouts/application.html.erb:
```diff
@@ -13,6 +13,7 @@
       <% flash.each do |key, value| %>
         <div class="alert alert-<%= key %>"><%= value %></div>
       <% end %>
+      <%= render partial: 'shared/user_bar' %>
```
I believe you dont need to specify partial here.

from inside a view file rails automatically assumes it will be a partial.

<%= render shared/user_bar' %>

qr8r reviewed (approved):
Much better PR! 

One tiny comment, but its nto enough to ask for changes here
